helix-term/src/commands.rs

```diff
@@ -3280,14 +3286,41 @@ pub mod insert {
     }
 
     pub fn insert_tab(cx: &mut Context) {
+        let (view, doc) = current_ref!(cx.editor);
+        let view_id = view.id;
+
+        if let Some(ref cmd) = cx.editor.config().supertab {
+            let cursors_after_whitespace = doc.selection(view_id).ranges().iter().all(|range| {
+                let cursor = range.cursor(doc.text().slice(..));
+                let current_line_num = doc.text().char_to_line(cursor);
+                let current_line_start = doc.text().line_to_char(current_line_num);
+                let left = doc.text().slice(current_line_start..cursor);
+                log::debug!("left: {:?}", left);
+                left.chars().all(|c| c.is_whitespace())
+            });
+
+            if !cursors_after_whitespace {
+                let cmd = match MappableCommand::from_str(cmd) {
```

Comment:
U think reparsing the mappable command everytime it's used is not ideal. You can do this during desexualization instead. In fact `MappableCommand` likely already implements desieralize (I didn't check so if it doesn't you can easily add an implementation, that just deserialzes a String and then uses the existing parse implementation).

Reply:
Yeah I think I tried this first, but the problem is `MappableCommand` is in `helix-term` while this config is in `helix-view`. I don't like parsing it every time either, but I'm not really sure how else to do it. I'm open to suggestions.

Reply:
ah the classic problem "commands are in helix-term but I need to use the in helix-view" :D

Hmm maybe since this is more of a special keybinding it could be a special kind of keybinding (the keybinding config lives in `helix-term`).

Reply:
Hm, so I took a look at the key map config:

https://github.com/helix-editor/helix/blob/53f47bc47771c94dab51626ca025be28e62eba0c/helix-term/src/config.rs#L13-L17

As we can see, it is a hash map that maps `Mode`s onto key maps. If we want this config somewhere in there, then I think we'd have two options:

1. Put it in an actual key map, as in

    ```toml
    [keys.insert]
    tab = # do insert tab or "move_parent_node_end"
    ```

    It's not really obvious to me here how to express what we want, without a generalized mechanism to detect any generic situation where you'd want to conditionally do one command or another, depending on the state of the text and the selection.

    I had actually considered the above idea when I first started thinking about this feature, but it seemed overly generic and complicated.
2. Put it one higher up under `keys`

    ```toml
    keys.supertab = "move_parent_node_end"
    ```

    But this doesn't map very well to the current config, which maps modes onto key maps.

Did you have any particular format in mind for moving into a key map config?

Also, I was thinking, currently `MappableCommand::from_str` is doing a linear search through every command. What if instead of the above, I made a static map of the commands, similar to how a map is made for `TypableCommand`?

https://github.com/helix-editor/helix/blob/d3cca0f3571fd30f4d7067a82c02753686297a67/helix-term/src/commands/typed.rs#L2809-L2818

This would get parsing down to O(1). Still not as good as no parsing at all, but significantly better than a needles linear search.

Reply:
I was actually thinking of
```toml
keymap.supertab = ...
```

When I made that comment. I agree that a tab mapping doesn't make sense since this is a very special cased feature and would feel weird to give special meaning for tab here.


However, I don't think `keymap.<mode>` colliding with `keymap.insert` is that weird. If you read a config file:


``` toml

[keymap]
supertab = "...

[keymap.normal]
...

```

It actually looks kind of.intuitive to first have some normal settings and then the various mode mappings and should be pretty easy to document too.


Improving the parsing speed with a lut would definitly be nice but I would.prefer to just avoid the parsing altogether. It's not a huge performance overhead but it just feels a bit hacky to me. And I would personally actually expect a supertab setting in `keymap` and not in `editor`

Reply:
Ok, I think I agree with all your points. The issue then becomes how to represent the config. Currently it's just a `HashMap<Mode, Keymap>`, so if we want to be able to have regular configuration in there too, then we have two options:

1. Reorganize key mapping config under another common key, e.g.

    ```toml
    [keys]
    supertab = "..."

    [keys.bindings.insert]
    ...
    ```
2. Write a custom parser that special cases the mode strings and throws everything else into the general config.

1 would be a large breaking change, but 2 feels a bit awkward. It's probably the way to go though. @archseer do you want to weigh in before I make the change?

Reply:
I would avoid a breaking change here since this would affect a lot of users. It looks nicer but we should reserve stuff like that for when we move to a script based config. 

Serde has the [flatten attribute](https://serde.rs/field-attrs.html#flatten) exactly for this usecase so it should be pretty straightforward to implement:

``` rust
#[derive(Deserialize)]
struct KeymapConfig {
   supertab: MappableCommand,
   #[serde(flatten)]
   modes: HashMap<String, MappableCommand>
}
```

Reply:
Oh neat! I didn't know about that attribute, that's a lot simpler than I was thinking. Thanks Pascal, I'll take that route

Reply:
Ok, I managed to switch this to put the `supertab` config into the `keys` section. It turned out to be more complicated than just using the `#[serde(flatten)]` attribute: I had to add the new keymap config object to a couple of context objects, but this may end up being useful in the future anyway, if we ever add more to this config section.

Let me know what you think.

Reply:
Ah I didn't think of that. It's not that much extra code tough and might be useful in the future too so I think that's ok 👍